### PR TITLE
Move MSBuild projects out of the native build scripts

### PR DESCRIPTION
**Pull request by jkoritzinsky**
### Summary

Significantly simplify our native build scripts by moving all of the managed builds up to the Arcade build infrastructure and splitting them into subsets to help save developer time on their inner builds.

Additionally, since it isn't easily possible to make Arcade build a project as self-contained publish, this PR also changes the crossgen2 build to build framework-dependent, have the Helix SDK pull down a .NET CLI for the test runs, and update the test run scripts to resolve the local dotnet CLI or the one in the path based on environment variables.

### Subsets

The coreclr build is split into these subsets:

- corelib
  - Contains System.Private.CoreLib
- runtime
  - The native build as well as crossgenning of System.Private.CoreLib
- linuxdac (Windows-only)
  - The cross-OS Windows->Linux DAC
- alpinedac (Windows-only)
  - The cross-OS Windows->Alpine Linux DAC
- nativecorelib
  - Crossgens System.Private.Corelib
- tools
  - The managed tools such as R2RDump and Crossgen2
- packages
  - The coreclr packages

### Local build improvements

The managed eventing headers are generated during the System.Private.CoreLib build by discovering python and running the script during the build in MSBuild.

The ucrt copy and enforce-pgo steps run as part of the native build instead of the managed build, which makes significantly more sense logically.

Due to keeping all of the managed builds within a single MSBuild invocation instead of nested invocations, this change saves around a minute in local build times.

### Workflow changes

CoreCLR developers will have to use the root `build.cmd/sh` scripts to build any of the managed components. The `src/coreclr/build-rutnime.cmd/sh` scripts only support building the native coreclr CMake project. The `src/coreclr/crossgen-corelib.cmd/sh` scripts support running crossgen on System.Private.CoreLib. The old `src/coreclr/build.cmd/sh` scripts are kept so as to give developers a transition period from the scripts to the new scripts.

#### Workflow examples

I've included some workflow examples below.

Build CoreCLR, System.Private.CoreLib, crossgen System.Private.CoreLib, managed tools, CoreCLR packages in Checked configuration for x86:

`D:\source\runtime>./build.cmd -subsetCategory coreclr -c checked -arch x86`

Build CoreCLR, System.Private.CoreLib, crossgen System.Private.CoreLib, skip managed tools, skip coreclr packages:

`D:\source\runtime>./build.cmd -subsetcategory coreclr -subset runtime-corelib-nativecorelib`

Build CoreCLR and managed tools only:

`D:\source\runtime>./build.cmd -subsetcategory coreclr -subset runtime-tools`

Build CoreCLR, System.Private.CoreLib, crossgen System.Private.CoreLib, managed tools, CoreCLR packages in debug and libraries in release (already works today):

`D:\source\runtime> ./build.cmd -subsetcategory coreclr-libraries -c Release -runtimeConfiguration debug`

### Future improvements

This PR will enable us to build System.Private.CoreLib in a job before running the coreclr or libraries builds. As a result, this moves us closer to being able to build coreclr and libraries in parallel in CI. Additionally, this does most of the work to Arcadify coreclr's build, so we can focus more on what our next steps and "final" build script design will be.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
>To help make this more visible, I'm thinking about renaming the build.cmd/sh scripts to be build-native.cmd/sh and splitting out the crossgen script to be more explicit.

This renaming is a good idea that I strongly suggest. Instead of completely removing the `build.cmd/sh` scripts I would instead leave them intact for some amount of time (2 - 3 weeks) and have them print out the new workflow (i.e. point to a readme/URL/PR). The breadcrumb of what to do next would be very helpful.

File: eng/build.sh
Comment (am11) [inline eng/build.sh]: Could we factor out CpuName and OSName blocks from https://github.com/dotnet/runtime/blob/bc5e7e36cefd1b53b654419cfd2bffc6b65dd403/eng/native/build-commons.sh#L240 in a separate file (something like `eng/native/initOSAndArch.sh`) and reuse? It used to be in four places and recently converged in a common script.

**Comment by jkoritzinsky** [comment]
Failures are https://github.com/dotnet/runtime/issues/32126

**Comment by jkoritzinsky** [comment]
@fadimounir can you test out a crossgen2 package from this PR to make sure it lays everything out correctly for the SDK to consume it?

File: src/coreclr/build.cmd
Comment (jaredpar) [inline src/coreclr/build.cmd]: Please file an issue that tracks the deletion of the sub-directory build scripts and add a link above this line. This way we broadcast our intent here and give contributors a place to discuss the issue.

File: src/coreclr/build.sh
Comment (jaredpar) [inline src/coreclr/build.sh]: Please file an issue that tracks the deletion of the sub-directory build scripts and add a link above this line. This way we broadcast our intent here and give contributors a place to discuss the issue.

File: src/coreclr/build-runtime.sh
Comment (jaredpar) [inline src/coreclr/build-runtime.sh]: Is this script meant for direct developer usage as say `build.cmd` formerly was?

File: src/coreclr/clr.featuredefines.props
Comment (jaredpar) [inline src/coreclr/clr.featuredefines.props]: Why is this the case? Can't see how VS would or would not interfere with this definition.

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj
Comment (jaredpar) [inline src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: The rest of this project file seems to prefer \ over /

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj
Comment (jaredpar) [inline src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: It feels a bit off to add a bunch of targets to the CoreCLR  build to discover the python executable.

Did you consider putting them in a more central location and then using a property, like `<RequiresPython>true</RequiresPython>` to allow projects to opt into using it? That would make it more generally useful and clean up this project file.

File: eng/native/init-os-and-arch.sh
Comment (lpereira) [inline eng/native/init-os-and-arch.sh]: Nitpicky suggestion:

```sh
case "$OSName" in
FreeBSD|Linux|NetBSD|OpenBSD|SunOS)
    os=$OSName ;;
Darwin)
	os=OSX ;;
*)
    echo "Unsupported OS $OSName detected, configuring as if for Linux"
    os=Linux ;;
esac
```

File: eng/native/init-os-and-arch.sh
Comment (lpereira) [inline eng/native/init-os-and-arch.sh]: Might want to case on `i[3-6]86` here instead, to be thorough.  (Even though machines older than i686 are *really* unlikely these days.)

File: eng/native/init-os-and-arch.sh
Comment (lpereira) [inline eng/native/init-os-and-arch.sh]: Maybe  case on `amd64|x86_64)` and get rid of the case below?

File: src/coreclr/clr.featuredefines.props
Comment (jkoritzinsky) [reply src/coreclr/clr.featuredefines.props]: This is outdated from a previous revision. I need to go remove this.

File: src/coreclr/build-runtime.sh
Comment (jkoritzinsky) [reply src/coreclr/build-runtime.sh]: Yes. For people who want to avoid using MSBuild as much as possible or for people who are building on platforms where we don't have a working MSBuild, they can call this script directly. It should have the same user experience as the old build.cmd/sh scripts except it only builds the CoreCLR runtime and cross components.

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj
Comment (jkoritzinsky) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: I'll go move them up so more projects can opt-in.

Comment (jkoritzinsky) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Now any target can depend on the `FindPython` target and python will be available for their scripts to use.

**Comment by fadimounir** [comment]
>  can you test out a crossgen2 package from this PR to make sure it lays everything out correctly for the SDK to consume it?

@jkoritzinsky where can I find the package? in which artifact is it?

**Comment by jkoritzinsky** [comment]
It looks like we don't publish the packages as artifacts in PR builds. Looks like we'll have to grab a local build. I can put one together for you and send it over.

**Comment by jkoritzinsky** [comment]
I've confirmed offline with Fadi that the package produced by the most recent commit in this PR works for the Crossgen2 SDK work.

File: src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj
Comment (am11) [inline src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: Is it missing the macOS RID? Would it be possible to include all supported RID from the repos RID graph file here and in readytorun's csproj?

The reason why I am asking is; while working on Android (#32800), I found that this csproj and readytorun csproj were failing to build because of the missing RID. If there is no dependency that is preventing to use all-rids graph in live-live build scenario, please consider adding those. :)

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: This matches the rids specified in the `src/installer/pkg/projects/netcoreapp/pkg/crossgenRIDs.props` file to enable the live-live pipeline to correctly generate a project.assets.json that can be used to create a self-contained deps.json for the crossgen2 package. Without this, the live-build pipeline was failing. I couldn't repro the failure locally, but this logically fixed the failure.

Comment (am11) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: Ah, I think one way is to Import that prop file at the beginning (after the `<Project ...>` line).

I was actually referring to all supported platform RIDs in: https://github.com/dotnet/runtime/blob/bd98efe/src/libraries/pkg/Microsoft.NETCore.Platforms/runtime.json.  If there is no technical blocker, could crossgen2 and readytorun csprojs use that list without whitelisting just a handful of RIDs, like rest/most of the csprojs in the runtime repo?

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: I'll see if I can finagle it to have a shared list.

Crossgen2 has a native dependency (the JIT), so it'd have to be updated to be bundled with the correct JIT, which it currently only does for Windows and Linux. It'd probably get macOS x64 correct, but it definitely isn't set up to set up cross-builds correctly.

I believe the list above is the supported list of RIDs for the .NET 5 timeframe anyway.

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: What was the particular failure you were seeing in #32800 for these projects? If it was the self-contained publish, that wouldn't be an issue here since the way the project is restored it would still work if we're building for a platform whose RID isn't in the list.

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: The more I think about it, I don't think we can centralize the list of RIDs until the crossgen2 team decides how they're going to support the distros we cross-build for. For this PR I'll keep the two lists separate unless someone from that team tells me otherwise.

Comment (am11) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: > What was the particular failure you were seeing

The error was emerging from `src/coreclr/build.sh`, where it calls `dotnet publish` on `crossgen2.csproj` (after compiling native components and building managed components successfully):
```
 error : NETSDK1056: Project is targeting runtime 'android.28-arm64' but did not resolve any runtime-specific packages. This runtime may not be supported by the target framework.
```

maybe we can skip building this project to get full E2E clean build on other platforms. (I haven't tested the behavior with your branch, perhaps it is already like that with the deprecation of `coreclr/build.sh`)

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: Sounds like it might be a publish-only issue. The new scripts don’t run dotnet publish (which is why this specific line I added needed to be added). I’m pretty sure that won’t be a problem, although it’d be good to check it out if you have a chance.

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj
Comment (jkoritzinsky) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: It looks like it's split across items and properties. Items use `\` whereas properties in this file tend to use `/`.

**Comment by jkoritzinsky** [comment]
@dotnet/runtime-infrastructure can I get another review pass on this PR?

**Comment by jaredpar** [comment]
Reviewing

**Comment by jaredpar** [comment]
Before merging we need to do an official signed build of this change. Too much of the core build logic is changing. Need to validate ahead of time it doesn't break official builds.

File: eng/python.targets
Comment (jaredpar) [inline eng/python.targets]: Our build requires python 3 so why are we using `py -2` and `python2`? That's basically allowing the build to proceed when we know it's going to fail. Better to error early and push developers to supported toolsets.

File: eng/Subsets.props
Comment (jkoritzinsky) [reply eng/Subsets.props]: This is the Runtime subset within the CoreCLR subset category. Mono could change their build to have a Runtime subset and then running `./build -subset runtime` would build both native runtimes.

File: eng/Subsets.props
Comment (jkoritzinsky) [reply eng/Subsets.props]: Mono's default doesn't include CoreCLR. Mono's default includes the "all" subset in the Mono subset category. Same way that Libraries' default includes the "all" subset in the Libraries subset category.

File: eng/python.targets
Comment (jkoritzinsky) [reply eng/python.targets]: I initially just pulled this from the old CoreCLR build script. I'll update it to only grab python3 installations.

File: eng/Subsets.props
Comment (jkoritzinsky) [reply eng/Subsets.props]: This is done to match the above lines for `CoreClrProjectToBuild` to be consistent.

File: src/coreclr/build-runtime.cmd
Comment (hoyosjs) [inline src/coreclr/build-runtime.cmd]: where are our built packages used in native bits?

**Review by hoyosjs — APPROVED** [review]
Except the little breaking comments

**Comment by hoyosjs** [comment]
> > To help make this more visible, I'm thinking about renaming the build.cmd/sh scripts to be build-native.cmd/sh and splitting out the crossgen script to be more explicit.
> 
> This renaming is a good idea that I strongly suggest. Instead of completely removing the `build.cmd/sh` scripts I would instead leave them intact for some amount of time (2 - 3 weeks) and have them print out the new workflow (i.e. point to a readme/URL/PR). The breadcrumb of what to do next would be very helpful.

Another thing: given the verbosity of these scripts, can we put the message as the end of the build?

File: src/coreclr/src/scripts/check-definitions.py
Comment (jkoritzinsky) [reply src/coreclr/src/scripts/check-definitions.py]: The definition consistency checking never really worked. It’s had warnings for months and no one could see them since we moved it into Arcade. It also doesn’t work with CMake Generator Expressions. @janvorli said it was fine if we removed it.

File: src/coreclr/build-runtime.cmd
Comment (hoyosjs) [inline src/coreclr/build-runtime.cmd]:
```suggestion
set "__LogsDir=%__RootBinDir%\log\!__BuildType!"
set "__MsbuildDebugLogsDir=%__LogsDir%\MsbuildDebugLogs"
```
Although it needs to be further down around __CrossCompIntermediatesDir

File: eng/Subsets.props
Comment (jaredpar) [reply eng/Subsets.props]: Seems confusing to name it "all" then not have all the items in it. Is there some subtle, or less subtle, context I'm missing here?

Comment (jkoritzinsky) [reply eng/Subsets.props]: The “all” refers to “all projects that are in this subset category” not “all projects in dotnet/runtime”.

File: eng/Subsets.props
Comment (jkoritzinsky) [reply eng/Subsets.props]: The parameter in the current scripts to skip it is skipnativecorlib I believe, so I tried to match the original naming design.

File: eng/build.ps1
Comment (jkoritzinsky) [reply eng/build.ps1]: It was only getting set during the CoreCLR build in the old scripts.

File: src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj
Comment (am11) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: I tested the Android changes by rebasing against your branch. CoreCLR build succeeded (native and managed components), no more errors about crossgen2. 👍  Libraries build continues to succeed (native and managed components).

However, in installer build, it has progressed to fully build native (with correct RID) and managed components (with wrong RID), thereby packaging step fails because of the wrong RID was picked up during the ngen and perfmap generation steps:

```sh
$ ROOTFS_DIR=$(realpath .tools/android-rootfs/android-ndk-r21/sysroot) ./build.sh --cross --arch arm64 --subsetCategory installer
__DistroRid: android.28-arm64
__RuntimeId: android.28-arm64
  Tool 'coverlet.console' (version '1.7.0') was restored. Available commands: coverlet
  Tool 'dotnet-reportgenerator-globaltool' (version '4.4.2') was restored. Available commands: reportgenerator
  Tools restore was successful.
  Restore completed in 85.39 ms for /datadrive/home/.nuget/packages/microsoft.dotnet.arcade.sdk/5.0.0-beta.20124.3/tools/Tools.proj.
[snip]
  __DistroRid: android.28-arm64  <--- these guys show up twice with this branch?
  __RuntimeId: android.28-arm64
[snip]
  Commencing build of "installer component" for Linux.arm64.Debug in /datadrive/projects/runtime4/artifacts/obj/android.28-arm64.Debug
[snip]
 [100%] Linking CXX executable nativehost
  [100%] Built target nativehost
  Install the project...
  -- Install configuration: "DEBUG"
  -- Installing: /datadrive/projects/runtime4/artifacts/bin/android.28-arm64.Debug/corehost/apphost
  -- Installing: /datadrive/projects/runtime4/artifacts/bin/android.28-arm64.Debug/corehost/dotnet
[snip]   <!--  interesting part: note the RID in path a line above and below  -!>
  Native image /datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.CSharp.dll generated successfully.
  Native image /datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.VisualBasic.Core.dll generated successfully.
[snip] <!- build continues with wrong RID -!>
  Successfully generated perfmap for native assembly '/datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.CSharp.dll'.
  Successfully generated perfmap for native assembly '/datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.VisualBasic.Core.dll'.
[snip] <!- packaging step fails because RID of native (and manged) components of coreclr, libraries are different than what installer's ngen step picks up -!>
  Microsoft.NETCore.DotNetAppHost -> /datadrive/projects/runtime4/artifacts/packages/Debug/specs/Microsoft.NETCore.DotNetAppHost.nuspec
/datadrive/home/.nuget/packages/microsoft.dotnet.build.tasks.packaging/5.0.0-beta.20124.3/build/Packaging.targets(1260,5): error : Error when creating nuget lib package from /datadrive/projects/runtime4/artifacts/packages/Debug/specs/runtime.linux-arm64.Microsoft.NETCore.DotNetAppHost.nuspec. System.IO.DirectoryNotFoundException: Could not find a part of the path '/datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/corehost'. [/datadrive/projects/runtime4/src/installer/pkg/projects/Microsoft.NETCore.DotNetAppHost/Microsoft.NETCore.DotNetAppHost.pkgproj]
/datadrive/home/.nuget/packages/microsoft.dotnet.build.tasks.packaging/5.0.0-beta.20124.3/build/Packaging.targets(1260,5): error :    at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound) [/datadrive/projects/runtime4/src/installer/pkg/projects/Microsoft.NETCore.DotNetAppHost/Microsoft.NETCore.DotNetAppHost.pkgproj]
/datadrive/home/.nuget/packages/microsoft.dotnet.build.tasks.packaging/5.0.0-beta.20124.3/build/Packaging.targets(1260,5): error :    at System.IO.Enumeration.FileSystemEnumerator`1.Init() [/datadrive/projects/runtime4/src/installer/pkg/projects/Microsoft.NETCore.DotNetAppHost/Microsoft.NETCore.DotNetAppHost.pkgproj]
```

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: I'll file a new issue on Monday to track rid unification throughout the repo.

Comment (am11) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: Thanks!
A nit, this segment:
```
__DistroRid: android.28-arm64
__RuntimeId: android.28-arm64
```
appears twice in build out with PR changes.

File: eng/pipelines/coreclr/templates/build-job.yml
Comment (jkoritzinsky) [reply eng/pipelines/coreclr/templates/build-job.yml]: Yep. And it looks like I might be able to simplify this based on that. I'll go clean this up.

File: Directory.Build.targets
Comment (ViktorHofer) [inline Directory.Build.targets]: nit: remove the slash as `RepositoryEngineeringDir` already has it in it: https://github.com/dotnet/runtime/blob/master/Directory.Build.props#L24 and https://github.com/dotnet/arcade/blob/e0112fe2f1fd7f6db08b228db98f22225f82267b/src/Microsoft.DotNet.Arcade.Sdk/tools/RepoLayout.props#L44

File: eng/build.ps1
Comment (ViktorHofer) [inline eng/build.ps1]: nit: use the arch variable in the subsequent statements in the same line.

File: eng/run-test.sh
Comment (ViktorHofer) [inline eng/run-test.sh]: Even though this is just a comment I would revert this back to `Configuration`. We don't use `ConfigurationGroup` anymore.

**Comment by jkoritzinsky** [comment]
There's no single exit point for src/coreclr/build.sh, so I'm just going to put the warning at the beginning and at the "success" end.

File: src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj
Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: That's expected currently since we run `init_distro_rid` twice, once in the root script and once in the coreclr-specific script. It's something we can clean up after this PR when we consolidate the RID discovery.

Comment (jkoritzinsky) [reply src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj]: Filed #33069 to track unifying the RID discovery methods

**Comment by jkoritzinsky** [comment]
Official build run at: https://dev.azure.com/dnceng/internal/_build/results?buildId=543897&view=results

**Comment by jkoritzinsky** [comment]
Yep. We're passing it as an "unprocessed build arg", which gets passed along to the MSBuild invocation for GenerateNativeVersionFile. If you check the official build I linked to, the outputs of the CoreCLR native build have the correct version info.

**Comment by jkoritzinsky** [comment]
Only failure is due to https://github.com/dotnet/runtime/issues/32764

Anyone else have any feedback? If not, I'm going to merge this at EOD.

**Pull request merged**
